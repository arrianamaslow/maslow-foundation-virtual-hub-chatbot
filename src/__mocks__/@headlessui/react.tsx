/**
 * @jest-environment jsdom
 */

import { DialogProps } from '@headlessui/react'
import React, { ReactNode } from 'react'

type DialogPanelTestProps = {
  children?: ReactNode
}

interface DialogTestProps extends DialogProps {
  children?: ReactNode
}

const DialogPanel: React.FC<DialogPanelTestProps> = ({ children }) => {
  return <div data-testid='mock-dialog-panel-component'>{children}</div>
}

const Dialog: React.FC<DialogTestProps> = ({ open, children }) => {
  return (
    <div data-testid='mock-dialog-component'>
      {open ? 'DIALOGUEOPENED' : 'DIALOGUECLOSED'}
      {children}
    </div>
  )
}

export { Dialog, DialogPanel }
