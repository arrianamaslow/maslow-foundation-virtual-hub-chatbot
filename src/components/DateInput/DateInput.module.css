.dateInputsContainer {
  @apply flex gap-[22px];
}

.errorMessageText {
  @apply text-[#FF0000] text-[1rem];
}
