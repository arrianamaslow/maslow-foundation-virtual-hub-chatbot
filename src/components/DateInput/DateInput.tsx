import React, { useState } from 'react'
import style from './DateInput.module.css'

export type DateInputProps = {
  setDate: (newDate: string, isValid: boolean) => void
  questionText: string
}

type dateState = {
  day: string
  month: string
  year: string
}

type dateValueError = {
  day: boolean
  month: boolean
  year: boolean
  wholeDate: boolean
}

const DateInput: React.FC<DateInputProps> = ({ setDate, questionText }) => {
  const [dateSelected, setDateSelected] = useState<dateState>({ day: '', month: '', year: '' })
  const [previousDateSelected, setPreviousDateSelected] = useState<string>()
  const [dateSelectedValueError, setDateSelectedValueError] = useState<dateValueError>({
    day: false,
    month: false,
    year: false,
    wholeDate: false
  })

  const isValidNumberBetweenRangeInclusive = (value: string, min: number, max: number) => {
    if (!/^\d+$/.test(value)) return false
    const integer = parseInt(value)
    if (integer < min || integer > max) return false
    return true
  }

  const checkDateValid = () => {
    // do not validate until all fields have a value
    if (!dateSelected.day || !dateSelected.month || !dateSelected.year) return

    const monthRangeInvalid = !isValidNumberBetweenRangeInclusive(dateSelected.month, 1, 12)

    const yearRangeInvalid = !isValidNumberBetweenRangeInclusive(
      dateSelected.year,
      1900,
      new Date().getFullYear()
    )

    const dayRangeInvalid = !isValidNumberBetweenRangeInclusive(dateSelected.day, 1, 31)

    let wholeDateInvalid = false

    const monthSelected = parseInt(dateSelected.month)
    const yearSelected = parseInt(dateSelected.year)
    const daySelected = parseInt(dateSelected.day)
    if (!dayRangeInvalid && !monthRangeInvalid && !yearRangeInvalid) {
      const validatedDate = new Date(yearSelected, monthSelected - 1, daySelected)
      if (
        validatedDate.getDate() !== daySelected ||
        validatedDate.getMonth() + 1 !== monthSelected ||
        validatedDate.getFullYear() !== yearSelected
      ) {
        wholeDateInvalid = true
      }
    }
    setDateSelectedValueError({
      day: dayRangeInvalid,
      month: monthRangeInvalid,
      year: yearRangeInvalid,
      wholeDate: wholeDateInvalid
    })

    const newDateSelected = `${yearSelected}-${monthSelected}-${daySelected}`
    if (!dayRangeInvalid && !monthRangeInvalid && !yearRangeInvalid && !wholeDateInvalid) {
      setDate(newDateSelected, true)
      setPreviousDateSelected(newDateSelected)
    } else {
      setDate(newDateSelected, false)
    }
  }

  return (
    <div className='md:h-[77px] h-[72px]'>
      <div className='text-left md:text-[1.313rem] text-[17px] font-normal font-lato'>
        {questionText}
      </div>
      <div className={style.dateInputsContainer}>
        <input
          data-testid='day'
          type='text'
          name={'day'}
          className={`w-[69px] text-center pr-0 pl-0  ${dateSelectedValueError.day && '!border-[#FF0000]'}`}
          placeholder='DD'
          maxLength={2}
          onChange={(e) => {
            setDateSelected({ ...dateSelected, day: e.currentTarget.value })
          }}
          onBlur={checkDateValid}
        />
        <input
          data-testid='month'
          type='text'
          name={'month'}
          className={`w-[69px] text-center pr-0 pl-0 ${dateSelectedValueError.month && '!border-[#FF0000]'}`}
          maxLength={2}
          placeholder='MM'
          onChange={(e) => {
            setDateSelected({ ...dateSelected, month: e.currentTarget.value })
          }}
          onBlur={checkDateValid}
        />
        <input
          data-testid='year'
          type='text'
          className={`w-[79px] text-center pr-0 pl-0 ${dateSelectedValueError.year && '!border-[#FF0000]'}`}
          name={'year'}
          placeholder='YYYY'
          maxLength={4}
          onChange={(e) => {
            setDateSelected({ ...dateSelected, year: e.currentTarget.value })
          }}
          onBlur={checkDateValid}
        />
      </div>
      {(dateSelectedValueError.day ||
        dateSelectedValueError.month ||
        dateSelectedValueError.year ||
        dateSelectedValueError.wholeDate) && (
        <div className={style.errorMessageText}>Must be a valid date</div>
      )}
    </div>
  )
}

export default DateInput
