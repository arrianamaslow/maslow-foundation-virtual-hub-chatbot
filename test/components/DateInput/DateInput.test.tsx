/**
 * @jest-environment jsdom
 */
import { render, screen, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'
import DateInput, { DateInputProps } from '@/components/DateInput/DateInput'

const defaultProps: DateInputProps = {
  setDate: () => {},
  questionText: ''
}

describe('<DateInput/>', () => {
  it('renders correctly', () => {
    const tree = render(<DateInput {...defaultProps} />)
    expect(tree).toMatchSnapshot()
  })

  it('has a day input field of type text, with placeholder text', () => {
    render(<DateInput {...defaultProps} />)
    const dayField = screen.getByTestId('day')
    expect(dayField).toHaveAttribute('type', 'text')
    expect(dayField).toHaveAttribute('placeholder', 'DD')
  })

  it('has a month input field of type text, with placeholder text', () => {
    render(<DateInput {...defaultProps} />)
    const monthField = screen.getByTestId('month')
    expect(monthField).toHaveAttribute('type', 'text')
    expect(monthField).toHaveAttribute('placeholder', 'MM')
  })

  it('has a year input field of type text with placeholder text', () => {
    render(<DateInput {...defaultProps} />)
    const yearField = screen.getByTestId('year')
    expect(yearField).toHaveAttribute('type', 'text')
    expect(yearField).toHaveAttribute('placeholder', 'YYYY')
  })

  it('runs the validation checker for the current value for the date, and calls the date update function where valid', () => {
    const dateSetter = jest.fn()
    render(<DateInput setDate={dateSetter} questionText='' />)

    const dayField = screen.getByTestId('day')
    fireEvent.change(dayField, { target: { value: '06' } })

    const monthField = screen.getByTestId('month')
    fireEvent.change(monthField, { target: { value: '08' } })

    const yearField = screen.getByTestId('year')
    fireEvent.change(yearField, { target: { value: '1991' } })

    fireEvent.blur(yearField)
    expect(dateSetter).toHaveBeenCalledWith('1991-8-6', true)
  })

  it('accepts a valid leap day as valid', () => {
    const dateSetter = jest.fn()
    render(<DateInput setDate={dateSetter} questionText='' />)

    const dayField = screen.getByTestId('day')
    fireEvent.change(dayField, { target: { value: '29' } })

    const monthField = screen.getByTestId('month')
    fireEvent.change(monthField, { target: { value: '02' } })

    const yearField = screen.getByTestId('year')
    fireEvent.change(yearField, { target: { value: '2024' } })

    fireEvent.blur(yearField)
    expect(dateSetter).toHaveBeenCalledWith('2024-2-29', true)
  })

  it('creates a div informing the user that their selection is invalid upon receiving an invalid date', () => {
    const dateSetter = jest.fn()
    render(<DateInput setDate={dateSetter} questionText='' />)

    const dayField = screen.getByTestId('day')
    fireEvent.change(dayField, { target: { value: 'this is not a valid date' } })

    const monthField = screen.getByTestId('month')
    fireEvent.change(monthField, { target: { value: 'this is not a valid date' } })

    const yearField = screen.getByTestId('year')
    fireEvent.change(yearField, { target: { value: 'this is not a valid date' } })

    fireEvent.blur(yearField)
    const complaintDiv = screen.getByText('Must be a valid date')
    expect(complaintDiv).not.toBeNull()
    expect(complaintDiv).toHaveClass('errorMessageText')
    expect(dateSetter).toHaveBeenCalledWith(expect.anything(), false)
  })

  it('does not accept an out of higher bound year', () => {
    const dateSetter = jest.fn()
    render(<DateInput setDate={dateSetter} questionText='' />)

    const dayField = screen.getByTestId('day')
    fireEvent.change(dayField, { target: { value: '06' } })

    const monthField = screen.getByTestId('month')
    fireEvent.change(monthField, { target: { value: '08' } })

    const yearField = screen.getByTestId('year')
    fireEvent.change(yearField, { target: { value: (new Date().getFullYear() + 1).toString() } })

    fireEvent.blur(yearField)
    const complaintDiv = screen.getByText('Must be a valid date')
    expect(complaintDiv).not.toBeNull()
    expect(complaintDiv).toHaveClass('errorMessageText')
    expect(dateSetter).toHaveBeenCalledWith(expect.anything(), false)
    expect(dayField).not.toHaveClass(`w-[69px] pr-0 !border-[#FF0000]`)
    expect(monthField).not.toHaveClass(`w-[69px] pr-0 !border-[#FF0000]`)
    expect(yearField).toHaveClass(`w-[79px] pr-0 !border-[#FF0000]`)
  })

  it('does not accept an out of lower bound year', () => {
    const dateSetter = jest.fn()
    render(<DateInput setDate={dateSetter} questionText='' />)

    const dayField = screen.getByTestId('day')
    fireEvent.change(dayField, { target: { value: '06' } })

    const monthField = screen.getByTestId('month')
    fireEvent.change(monthField, { target: { value: '08' } })

    const yearField = screen.getByTestId('year')
    fireEvent.change(yearField, { target: { value: '1899' } })

    fireEvent.blur(yearField)
    const complaintDiv = screen.getByText('Must be a valid date')
    expect(complaintDiv).not.toBeNull()
    expect(complaintDiv).toHaveClass('errorMessageText')
    expect(dateSetter).toHaveBeenCalledWith(expect.anything(), false)
    expect(dayField).not.toHaveClass(`w-[69px] pr-0 !border-[#FF0000]`)
    expect(monthField).not.toHaveClass(`w-[69px] pr-0 !border-[#FF0000]`)
    expect(yearField).toHaveClass(`w-[79px] pr-0 !border-[#FF0000]`)
  })

  it('does not accept an out of lower bound month', () => {
    const dateSetter = jest.fn()
    render(<DateInput setDate={dateSetter} questionText='' />)

    const dayField = screen.getByTestId('day')
    fireEvent.change(dayField, { target: { value: '06' } })

    const monthField = screen.getByTestId('month')
    fireEvent.change(monthField, { target: { value: '00' } })

    const yearField = screen.getByTestId('year')
    fireEvent.change(yearField, { target: { value: '2023' } })

    fireEvent.blur(yearField)
    const complaintDiv = screen.getByText('Must be a valid date')
    expect(complaintDiv).not.toBeNull()
    expect(complaintDiv).toHaveClass('errorMessageText')
    expect(dateSetter).toHaveBeenCalledWith(expect.anything(), false)
    expect(dayField).not.toHaveClass(`w-[69px] pr-0 !border-[#FF0000]`)
    expect(monthField).toHaveClass(`w-[69px] pr-0 !border-[#FF0000]`)
    expect(yearField).not.toHaveClass(`w-[79px] pr-0 !border-[#FF0000]`)
  })

  it('does not accept an out of higher bound month', () => {
    const dateSetter = jest.fn()
    render(<DateInput setDate={dateSetter} questionText='' />)

    const dayField = screen.getByTestId('day')
    fireEvent.change(dayField, { target: { value: '06' } })

    const monthField = screen.getByTestId('month')
    fireEvent.change(monthField, { target: { value: '13' } })

    const yearField = screen.getByTestId('year')
    fireEvent.change(yearField, { target: { value: '2023' } })

    fireEvent.blur(yearField)
    const complaintDiv = screen.getByText('Must be a valid date')
    expect(complaintDiv).not.toBeNull()
    expect(complaintDiv).toHaveClass('errorMessageText')
    expect(dateSetter).toHaveBeenCalledWith(expect.anything(), false)
    expect(dayField).not.toHaveClass(`w-[69px] pr-0 !border-[#FF0000]`)
    expect(monthField).toHaveClass(`w-[69px] pr-0 !border-[#FF0000]`)
    expect(yearField).not.toHaveClass(`w-[79px] pr-0 !border-[#FF0000]`)
  })

  it('does not accept an out of lower bound day', () => {
    const dateSetter = jest.fn()
    render(<DateInput setDate={dateSetter} questionText='' />)

    const dayField = screen.getByTestId('day')
    fireEvent.change(dayField, { target: { value: '00' } })

    const monthField = screen.getByTestId('month')
    fireEvent.change(monthField, { target: { value: '11' } })

    const yearField = screen.getByTestId('year')
    fireEvent.change(yearField, { target: { value: '2023' } })

    fireEvent.blur(yearField)
    const complaintDiv = screen.getByText('Must be a valid date')
    expect(complaintDiv).not.toBeNull()
    expect(complaintDiv).toHaveClass('errorMessageText')
    expect(dateSetter).toHaveBeenCalledWith(expect.anything(), false)
    expect(dayField).toHaveClass(`w-[69px] pr-0 !border-[#FF0000]`)
    expect(monthField).not.toHaveClass(`w-[69px] pr-0 !border-[#FF0000]`)
    expect(yearField).not.toHaveClass(`w-[79px] pr-0 !border-[#FF0000]`)
  })

  it('does not accept an out of higher bound day', () => {
    const dateSetter = jest.fn()
    render(<DateInput setDate={dateSetter} questionText='' />)

    const dayField = screen.getByTestId('day')
    fireEvent.change(dayField, { target: { value: '32' } })

    const monthField = screen.getByTestId('month')
    fireEvent.change(monthField, { target: { value: '11' } })

    const yearField = screen.getByTestId('year')
    fireEvent.change(yearField, { target: { value: '2023' } })

    fireEvent.blur(yearField)
    const complaintDiv = screen.getByText('Must be a valid date')
    expect(complaintDiv).not.toBeNull()
    expect(complaintDiv).toHaveClass('errorMessageText')
    expect(dateSetter).toHaveBeenCalledWith(expect.anything(), false)
    expect(dayField).toHaveClass(`w-[69px] pr-0 !border-[#FF0000]`)
    expect(monthField).not.toHaveClass(`w-[69px] pr-0 !border-[#FF0000]`)
    expect(yearField).not.toHaveClass(`w-[79px] pr-0 !border-[#FF0000]`)
  })

  it('does not accept a an invalid leap day', () => {
    const dateSetter = jest.fn()
    render(<DateInput setDate={dateSetter} questionText='' />)

    const dayField = screen.getByTestId('day')
    fireEvent.change(dayField, { target: { value: '29' } })

    const monthField = screen.getByTestId('month')
    fireEvent.change(monthField, { target: { value: '02' } })

    const yearField = screen.getByTestId('year')
    fireEvent.change(yearField, { target: { value: '2023' } })

    fireEvent.blur(yearField)
    const complaintDiv = screen.getByText('Must be a valid date')
    expect(complaintDiv).not.toBeNull()
    expect(complaintDiv).toHaveClass('errorMessageText')
    expect(dateSetter).toHaveBeenCalledWith(expect.anything(), false)
  })

  it('does not respond when day is missing', () => {
    const dateSetter = jest.fn()
    render(<DateInput setDate={dateSetter} questionText='' />)

    const monthField = screen.getByTestId('month')
    fireEvent.change(monthField, { target: { value: '08' } })

    const yearField = screen.getByTestId('year')
    fireEvent.change(yearField, { target: { value: '1991' } })

    fireEvent.blur(yearField)
    expect(dateSetter).not.toHaveBeenCalled()
    const complaintDiv = screen.queryByText('Must be a valid date')
    expect(complaintDiv).toBeNull()
  })

  it('does not respond when month is missing', () => {
    const dateSetter = jest.fn()
    render(<DateInput setDate={dateSetter} questionText='' />)

    const dayField = screen.getByTestId('day')
    fireEvent.change(dayField, { target: { value: '06' } })

    const yearField = screen.getByTestId('year')
    fireEvent.change(yearField, { target: { value: '1991' } })

    fireEvent.blur(yearField)
    expect(dateSetter).not.toHaveBeenCalledWith()
    const complaintDiv = screen.queryByText('Must be a valid date')
    expect(complaintDiv).toBeNull()
  })

  it('does not respond when year is missing', () => {
    const dateSetter = jest.fn()
    render(<DateInput setDate={dateSetter} questionText='' />)

    const dayField = screen.getByTestId('day')
    fireEvent.change(dayField, { target: { value: '06' } })

    const monthField = screen.getByTestId('month')
    fireEvent.change(monthField, { target: { value: '08' } })

    fireEvent.blur(monthField)
    expect(dateSetter).not.toHaveBeenCalled()
    const complaintDiv = screen.queryByText('Must be a valid date')
    expect(complaintDiv).toBeNull()
  })

  it('does not accept errors from multiple fields', () => {
    const dateSetter = jest.fn()
    render(<DateInput setDate={dateSetter} questionText='' />)

    const dayField = screen.getByTestId('day')
    fireEvent.change(dayField, { target: { value: '00' } })

    const monthField = screen.getByTestId('month')
    fireEvent.change(monthField, { target: { value: '13' } })

    const yearField = screen.getByTestId('year')
    fireEvent.change(yearField, { target: { value: '2023' } })

    fireEvent.blur(yearField)
    const complaintDiv = screen.getByText('Must be a valid date')
    expect(complaintDiv).not.toBeNull()
    expect(dateSetter).toHaveBeenCalledWith(expect.anything(), false)
    expect(dayField).toHaveClass(`w-[69px] pr-0 !border-[#FF0000]`)
    expect(monthField).toHaveClass(`w-[69px] pr-0 !border-[#FF0000]`)
    expect(yearField).not.toHaveClass(`w-[79px] pr-0 !border-[#FF0000]`)
  })
})
