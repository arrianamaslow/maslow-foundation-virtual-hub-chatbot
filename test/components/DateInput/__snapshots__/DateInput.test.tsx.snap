// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DateInput/> renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="md:h-[77px] h-[72px]"
      >
        <div
          class="text-left md:text-[1.313rem] text-[17px] font-normal font-lato"
        />
        <div
          class="dateInputsContainer"
        >
          <input
            class="w-[69px] text-center pr-0 pl-0  false"
            data-testid="day"
            maxlength="2"
            name="day"
            placeholder="DD"
            type="text"
          />
          <input
            class="w-[69px] text-center pr-0 pl-0 false"
            data-testid="month"
            maxlength="2"
            name="month"
            placeholder="MM"
            type="text"
          />
          <input
            class="w-[79px] text-center pr-0 pl-0 false"
            data-testid="year"
            maxlength="4"
            name="year"
            placeholder="YYYY"
            type="text"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="md:h-[77px] h-[72px]"
    >
      <div
        class="text-left md:text-[1.313rem] text-[17px] font-normal font-lato"
      />
      <div
        class="dateInputsContainer"
      >
        <input
          class="w-[69px] text-center pr-0 pl-0  false"
          data-testid="day"
          maxlength="2"
          name="day"
          placeholder="DD"
          type="text"
        />
        <input
          class="w-[69px] text-center pr-0 pl-0 false"
          data-testid="month"
          maxlength="2"
          name="month"
          placeholder="MM"
          type="text"
        />
        <input
          class="w-[79px] text-center pr-0 pl-0 false"
          data-testid="year"
          maxlength="4"
          name="year"
          placeholder="YYYY"
          type="text"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
